packages/frontend/src/contexts/AppContext/useTxHistory.ts

```diff
@@ -25,86 +27,200 @@ const cacheKey = 'recentTransactions:v000'
-        if (oldPending !== pending) {
-          updateTransaction(tx, { pending })
+
+      const clonedTxs = _.cloneDeepWith(prevTransactions, customizer)
+      const targetTx = find(clonedTxs, ['hash', matchingHash || tx.hash])
+
+      if (targetTx) {
+        for (const key in updateOpts) {
+          targetTx[key] = updateOpts[key]
         }
       }
-      const cbPendingDC = (pendingDestinationConfirmation: boolean) => {
-        if (oldPendingDC !== pendingDestinationConfirmation) {
-          updateTransaction(tx, { pendingDestinationConfirmation })
+      
+      return sortByRecentTimestamp(clonedTxs).slice(0, 4)
+    })
+  }
+
+  // stores hashes for either pending origin or destination confirmation to prevent redundant listeners
+  const listenerSet = useRef(new Set())
+
+  const listenForOriginConfirmation = async (tx) => {
+    try {
+      tx.provider.once(tx.hash, transaction => {
+        listenerSet.current.delete(tx.hash)
+        updateTransaction(tx, { pending: false })
+      })
+    } catch (e) {
+      console.error('Error with origin transaction listener:', e)
+      listenerSet.current.delete(tx.hash)
+    }
+  }
+
+  // stores setInterval IDs -- may be cleared to stop polling
+  const intervalRefs = useRef({})
+  // stores setTimeouts to limit polling to one hour
+  const timeoutRefs = useRef({})
+
+  const POLLING_INTERVAL = 15000
+  const POLLING_TIMEOUT = 3600000
+
+  const getBondedTxHash = (tx) => {
+    return new Promise((resolve, reject) => {
+      const network = (typeof process !== 'undefined' && process.env && process.env.REACT_APP_NETWORK) ? process.env.REACT_APP_NETWORK : 'goerli'
+      const explorerAPIUrl = `https://${network === 'goerli' && 'goerli-'}explorer-api.hop.exchange/v1/transfers?transferId=${tx.hash}`
+
+      if (!timeoutRefs.current[tx.hash]) {
+        timeoutRefs.current[tx.hash] = setTimeout(() => {
+          clearInterval(intervalRefs.current[tx.hash])
+          updateTransaction(tx, { pendingDestinationConfirmation: false })
+          reject(new Error('Polling timed out'))
+        }, POLLING_TIMEOUT)
+      }
+
+      const fetchAPI = async () => {
+        try {
+          const response = await fetch(explorerAPIUrl)
+          if (!response.ok) throw new Error('API request failed')
+
+          const responseJSON = await response.json()
+          if (!responseJSON.data || !responseJSON.data[0]) {
+            return
+          }
+
+          const bondTransactionHash = responseJSON.data[0].bondTransactionHash
+          if (bondTransactionHash) {
+            clearInterval(intervalRefs.current[tx.hash])
+            clearTimeout(timeoutRefs.current[tx.hash])
+            resolve(bondTransactionHash)
+          }
+        } catch (e) {
+          console.error(e)
         }
       }
-      tx.once('pending', cbPending)
-      tx.once('pendingDestinationConfirmation', cbPendingDC)
+
+      const interval = setInterval(fetchAPI, POLLING_INTERVAL)
```

Comment:
suggestion: use `useQuery` and use the `tx.hash` as the query key. It should prevent duplicated fetching